x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/template/IngestPipelineConfig.java

```diff
@@ -50,15 +54,21 @@ public int getVersion() {
         return version;
     }
 
-    public String getVersionProperty() {
-        return versionProperty;
-    }
-
     public List<String> getPipelineDependencies() {
         return dependencies;
     }
 
-    public BytesReference loadConfig() {
-        return new BytesArray(TemplateUtils.loadTemplate(resource, String.valueOf(version), versionProperty));
+    public PutPipelineRequest getPutPipelineRequest() {
+        return requestSupplier.get();
+    }
+
+    private static Supplier<PutPipelineRequest> newResourceSupplier(String id, String resource, int version, String versionProperty) {
+        Objects.requireNonNull(resource);
+        Objects.requireNonNull(versionProperty);
+        return () -> new PutPipelineRequest(
+            id,
+            new BytesArray(TemplateUtils.loadTemplate(resource, String.valueOf(version), versionProperty)),
+            XContentType.JSON
+        );
```

Comment:
Nice convenience method, but I think out of scope here. This is a dumb class that only describes how the ingest pipeline resource is read from a file into memory, it doesn't know anything about the cluster or what a request is. I'd maintain this limited responsibility and keep it similar to its counterparts that serve the same purpose for other resources loaded by registries.

Reply:
I've already paged out most of the details from my mind, but I think I added this so the pipeline resources didn't have to live in the x-pack-core plugin. Or maybe it was so the pipelines could be written as YAML. Or possibly both.

Reply:
We can probably achieve both by providing an APM-specific subclass of `IngestPipelineConfig` with a custom `loadConfig()` overload. I'll see how it looks like, it may be better in terms of keeping this class's responsibility to only load the resource from disk.